In [472]:
import pandas as pd

#### Pruning the edgelists 

In [473]:
total = pd.read_csv("../data/cleaned/total_edgelist.csv")
unique = pd.read_csv("../data/cleaned/unique_edgelist.csv")
repro = pd.read_csv("../data/cleaned/reproducible_edgelist.csv")
total_hyst = pd.read_csv("../data/cleaned/edgelist_total_hyst.csv")
unique_hyst = pd.read_csv("../data/cleaned/edgelist_unique_hyst.csv")
repro_hyst = pd.read_csv("../data/cleaned/edgelist_repro_hyst.csv")
distance = pd.read_csv("../data/cleaned/distance_edgelist.csv")
colony = pd.read_csv("../data/cleaned/colonizer_edgelist.csv")
comcol = pd.read_csv("../data/cleaned/comcol_edgelist.csv")
language = pd.read_csv("../data/cleaned/language_edgelist.csv")
contiguity = pd.read_csv("../data/cleaned/contiguity_edgelist.csv")
rta = pd.read_csv("../data/cleaned/rta_edgelist.csv")
attribute2000_2023 = pd.read_csv("../data/cleaned/attributes2000_2023.csv")

In [474]:
print(total['iso_o'].nunique())
print(unique['iso_o'].nunique())
print(repro['iso_o'].nunique())
print(total_hyst['iso_o'].nunique())
print(unique_hyst['iso_o'].nunique())
print(repro_hyst['iso_o'].nunique())
print(distance['iso_o'].nunique())
print(colony['iso_o'].nunique())
print(comcol['iso_o'].nunique())
print(language['iso_o'].nunique())
print(contiguity['iso_o'].nunique())
print(rta['iso_o'].nunique())
print(attribute2000_2023['iso_o'].nunique())

190
190
190
190
190
190
190
190
190
190
190
190
190


**I am keeping only the countries present in all edgelists and attribute datasets. Hence, I am conducting an inner join process.** 

The steps to prune the edgelists:
>1. Finding out the countries those are not in the attributes edgelist.
>2. Removing those countries from origin column ['iso_o'].
>3. Removing those countries from destination column ['iso_d'].
>4. Checking whether the countries in the origin and destination of an edgelist are same
>5. Cheking whether the number of countries in the origin and destination of an edgelist are same
>6. Replacing the original edgelist saved in the 'cleaned' subfolder of 'data' folder by saving it with the same name as 

* Total

In [475]:
set(total['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [476]:
total = total[(total['iso_o'] != 'MSR') & (total['iso_o'] != 'TWN')]
total = total[(total['iso_d'] != 'MSR') & (total['iso_d'] != 'TWN')]

In [477]:
set(total['iso_o'].unique()) - set(total['iso_o'].unique())

set()

In [478]:
print(total['iso_o'].nunique())
print(total['iso_d'].nunique())

190
190


In [479]:
total.to_csv("../data/cleaned/total_edgelist.csv", encoding = 'utf-8', index = False)

In [480]:
set(total_hyst['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [481]:
set(total['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [482]:
total_hyst = total_hyst[(total_hyst['iso_o'] != 'MSR') & (total_hyst['iso_o'] != 'TWN')]
total_hyst = total_hyst[(total_hyst['iso_d'] != 'MSR') & (total_hyst['iso_d'] != 'TWN')]

In [483]:
set(total_hyst['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [484]:
print(total_hyst['iso_o'].nunique())
print(total_hyst['iso_d'].nunique())

190
190


In [485]:
total_hyst.to_csv("../data/cleaned/edgelist_total_hyst.csv", encoding = 'utf-8', index = False)

* Unique

In [486]:
set(unique['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [487]:
unique = unique[(unique['iso_o'] != 'MSR') & (unique['iso_o'] != 'TWN')]
unique = unique[(unique['iso_d'] != 'MSR') & (unique['iso_d'] != 'TWN')]

In [488]:
set(unique['iso_o'].unique()) - set(unique['iso_o'].unique())

set()

In [489]:
print(unique['iso_o'].nunique())
print(unique['iso_d'].nunique())

190
190


In [490]:
unique.to_csv("../data/cleaned/unique_edgelist.csv", encoding = 'utf-8', index = False)

In [491]:
set(unique_hyst['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [492]:
unique_hyst = unique_hyst[(unique_hyst['iso_o'] != 'MSR') & (unique_hyst['iso_o'] != 'TWN')]
unique_hyst = unique_hyst[(unique_hyst['iso_d'] != 'MSR') & (unique_hyst['iso_d'] != 'TWN')]

In [493]:
set(unique_hyst['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [494]:
print(unique_hyst['iso_o'].nunique())
print(unique_hyst['iso_d'].nunique())

190
190


In [495]:
unique_hyst.to_csv("../data/cleaned/edgelist_unique_hyst.csv", encoding = 'utf-8', index = False)

* Reproducible

In [496]:
set(repro['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [497]:
repro = repro[(repro['iso_o'] != 'MSR') & (repro['iso_o'] != 'TWN')]
repro = repro[(repro['iso_d'] != 'MSR') & (repro['iso_d'] != 'TWN')]

In [498]:
set(repro['iso_o'].unique()) - set(unique['iso_o'].unique())

set()

In [499]:
print(repro['iso_o'].nunique())
print(repro['iso_d'].nunique())

190
190


In [500]:
repro.to_csv("../data/cleaned/reproducible_edgelist.csv", encoding = 'utf-8', index = False)

In [501]:
set(repro_hyst['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [502]:
repro_hyst = repro_hyst[(repro_hyst['iso_o'] != 'MSR') & (repro_hyst['iso_o'] != 'TWN')]
repro_hyst = repro_hyst[(repro_hyst['iso_d'] != 'MSR') & (repro_hyst['iso_d'] != 'TWN')]

In [503]:
set(repro_hyst['iso_o'].unique()) - set(unique['iso_o'].unique())

set()

In [504]:
print(repro_hyst['iso_o'].nunique())
print(repro_hyst['iso_d'].nunique())

190
190


In [505]:
repro_hyst.to_csv("../data/cleaned/edgelist_repro_hyst.csv", encoding = 'utf-8', index = False)

* Distance

In [506]:
set(distance['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [507]:
distance = distance[(distance['iso_o'] != 'MSR') & (distance['iso_o'] != 'TWN')]
distance = distance[(distance['iso_d'] != 'MSR') & (distance['iso_d'] != 'TWN')]

In [508]:
set(distance['iso_o'].unique()) - set(distance['iso_o'].unique())

set()

In [509]:
print(distance['iso_o'].nunique())
print(distance['iso_d'].nunique())

190
190


In [510]:
distance.to_csv("../data/cleaned/distance_edgelist.csv", encoding = 'utf-8', index = False)

* Contiguity

In [511]:
set(contiguity['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [512]:
contiguity = contiguity[(contiguity['iso_o'] != 'MSR') & (contiguity['iso_o'] != 'TWN')]
contiguity = contiguity[(contiguity['iso_d'] != 'MSR') & (contiguity['iso_d'] != 'TWN')]

In [513]:
set(contiguity['iso_o'].unique()) - set(contiguity['iso_o'].unique())

set()

In [514]:
print(contiguity['iso_o'].nunique())
print(contiguity['iso_d'].nunique())

190
190


In [515]:
contiguity.to_csv("../data/cleaned/contiguity_edgelist.csv", encoding = 'utf-8', index = False)

* Colonizer

In [516]:
set(colony['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [517]:
colony = colony[(colony['iso_o'] != 'MSR') & (colony['iso_o'] != 'TWN')]
colony = colony[(colony['iso_d'] != 'MSR') & (colony['iso_d'] != 'TWN')]

In [518]:
set(colony['iso_o'].unique()) - set(colony['iso_o'].unique())

set()

In [519]:
print(colony['iso_o'].nunique())
print(colony['iso_d'].nunique())

190
190


In [520]:
colony.to_csv("../data/cleaned/colonizer_edgelist.csv", encoding = 'utf-8', index = False)

* Common Colonizer

In [521]:
set(comcol['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [522]:
comcol = comcol[(comcol['iso_o'] != 'MSR') & (comcol['iso_o'] != 'TWN')]
comcol = comcol[(comcol['iso_d'] != 'MSR') & (comcol['iso_d'] != 'TWN')]

In [523]:
set(comcol['iso_o'].unique()) - set(comcol['iso_o'].unique())

set()

In [524]:
print(comcol['iso_o'].nunique())
print(comcol['iso_d'].nunique())

190
190


In [525]:
comcol.to_csv("../data/cleaned/comcol_edgelist.csv", encoding = 'utf-8', index = False)

* Language

In [526]:
set(language['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

set()

In [527]:
language = language[(language['iso_o'] != 'MSR') & (language['iso_o'] != 'TWN')]
language = language[(language['iso_d'] != 'MSR') & (language['iso_d'] != 'TWN')]

In [528]:
set(language['iso_o'].unique()) - set(language['iso_o'].unique())

set()

In [529]:
print(language['iso_o'].nunique())
print(language['iso_d'].nunique())

190
190


In [530]:
language.to_csv("../data/cleaned/language_edgelist.csv", encoding = 'utf-8', index = False)

* RTA

In [531]:
l = set(rta['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique())

In [532]:
len(l)

0

In [533]:
280 - 90

190

In [534]:
rta = rta[~rta['iso_o'].isin(l)]
rta = rta[~rta['iso_d'].isin(l)]

In [535]:
print(rta['iso_o'].nunique())
print(rta['iso_d'].nunique())

190
190


In [536]:
total['iso_o'].nunique()

190

In [537]:
rta.to_csv("../data/cleaned/rta_edgelist.csv", encoding = 'utf-8', index = False)

In [538]:
assert rta['iso_o'].nunique() == total['iso_o'].nunique() == unique['iso_o'].nunique() == repro['iso_o'].nunique() == distance['iso_o'].nunique() == comcol['iso_o'].nunique() == colony['iso_o'].nunique() == language['iso_o'].nunique() == contiguity['iso_o'].nunique() == attribute2000_2023['iso_o'].nunique() 

In [539]:
assert rta['iso_d'].nunique() == total['iso_d'].nunique() == unique['iso_d'].nunique() == repro['iso_d'].nunique() == distance['iso_d'].nunique() == comcol['iso_d'].nunique() == colony['iso_d'].nunique() == language['iso_d'].nunique() == contiguity['iso_d'].nunique()

In [540]:
assert set(rta['iso_d'].unique()) == set(total['iso_d'].unique()) == set(unique['iso_d'].unique()) == set(repro['iso_d'].unique()) == set(distance['iso_d'].unique()) == set(comcol['iso_d'].unique()) == set(colony['iso_d'].unique()) == set(language['iso_d'].unique()) == set(contiguity['iso_d'].unique())

In [541]:
assert set(rta['iso_o'].unique()) == set(total['iso_o'].unique()) == set(unique['iso_o'].unique()) == set(repro['iso_o'].unique()) == set(distance['iso_o'].unique()) == set(comcol['iso_o'].unique()) == set(colony['iso_o'].unique()) == set(language['iso_o'].unique()) == set(contiguity['iso_o'].unique())

In [542]:
print(total['iso_o'].nunique())
print(unique['iso_o'].nunique())
print(repro['iso_o'].nunique())
print(total_hyst['iso_o'].nunique())
print(unique_hyst['iso_o'].nunique())
print(repro_hyst['iso_o'].nunique())
print(distance['iso_o'].nunique())
print(comcol['iso_o'].nunique())
print(colony['iso_o'].nunique())
print(language['iso_o'].nunique())
print(contiguity['iso_o'].nunique())
print(rta['iso_o'].nunique())
print(attribute2000_2023['iso_o'].nunique())

190
190
190
190
190
190
190
190
190
190
190
190
190


### Preparing the panel data

**Steps to prepare the panel dataet:**
>1. Removing the year 1999 from attribute2000_2023 dataframe for removing the NaN columns of first differences variables
>2. Renaming the 'time' column to 'TIME_PERIOD" of attributes data to match with the other dataset
>3. Finding out and removing the countries those are present in the long format trade dataset but are not in the attributes dataset from both origin and destination
>4. Renaming the 'OBS_VALUE' column of the respective trade data frame as 'total_tr', 'unique_cultural_tr'and 'reproducible_cultural_tr'
>5. Finding out and removing the countries those are present in the dataset containing all fixed factors (both node and dyad level) but are not in the attributes dataset from both origin and destination
>6. Imputing missing values for the 'development' column of the fixed long data from WDI country classification based on income.
>7. Merging the attributes data with total trade data first as the attributes has highest rows for each observation ['iso_o'] and ordering the columns and naming it as 'panel_total'
>8. Removing 'PRODUCT_HS' and 'ADJUSTMENT' columns from unique cultural trade data and merging with the 'panel_total' on 'iso_o', 'iso_d' and 'TIME_PERIOD', using left joing method and naming the new data as 'panel_total2'
>9. Removing 'PRODUCT_HS' and 'product_count' columns from reproducible cultual trade data and merging with the 'panel_total2' on 'iso_o', 'iso_d' and 'TIME_PERIOD', using left joing method and naming the new data as 'panel_total3'
>10. From Itertools I made a dataframe of all possible country pair for all 24 years that have 190*190*24 rows.
>11. I removed the selfloops from that dataframe with three columns.
>12. I merged it with the subset of the panel_data3 dataframe created at step 7 on 'iso_o', 'iso_d' and 'TIME_PERIOD'
>13. Then I merged 'attribute2000_2023' which is the country level attribute data on 'iso_o' and 'TIME_PERIOD'
>14. Created a copy of the dataframe, renaming all the colum names with suffix '_d', removing the 'iso_o' column
>15. Removing duplicates on `['iso_d', 'TIME_PERIOD']`
>16. Checking that every country appears 24 times (once for every year)
>17. Merging this data on `['iso_d', 'TIME_PERIOD']`
>18. Merging fixed_long data (having node and dyad level fixed factors) with it on 'iso_o' and 'iso_d'
>10. Sanity check steps and saving as csv

In [543]:
attribute2000_2023 = attribute2000_2023.rename(columns = {'time': 'TIME_PERIOD'})

In [544]:
total_long = pd.read_csv("../data/cleaned/total2000_2023.csv")
unique_long = pd.read_csv("../data/cleaned/unique2000_2023.csv")
repro_long = pd.read_csv("../data/cleaned/reproducible2000_2023.csv")
fixed_long = pd.read_csv("../data/cleaned/fixed.csv")

In [545]:
print(attribute2000_2023['iso_o'].nunique())
print(fixed_long['iso_o'].nunique())
print(total_long['iso_o'].nunique())

190
224
201


In [546]:
224 - 190

34

In [547]:
print(set(total_long['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique()))
print(set(unique_long['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique()))
print(set(repro_long['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique()))
print(set(fixed_long['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique()))

{'PSE', 'MSR', 'WXD', 'SSD', 'TLS', 'ROU', 'TWN', 'SCG_F', 'SRB', 'MNE', 'COD'}
{'PSE', 'MSR', 'WXD', 'SSD', 'TLS', 'ROU', 'TWN', 'SCG_F', 'SRB', 'MNE', 'COD'}
{'PSE', 'MSR', 'WXD', 'SSD', 'TLS', 'ROU', 'TWN', 'SCG_F', 'SRB', 'MNE', 'COD'}
{'TCA', 'TWN', 'ESH', 'GIB', 'MSR', 'MTQ', 'CCK', 'VGB', 'SOM', 'WLF', 'ANT', 'MNP', 'SHN', 'YUG', 'FRO', 'AIA', 'NFK', 'PCN', 'PAL', 'CYM', 'SPM', 'ROM', 'FLK', 'TKL', 'COK', 'PRI', 'CXR', 'NIU', 'GLP', 'GUF', 'REU', 'SMR', 'TMP', 'ZAR'}


In [548]:
 attribute2000_2023.head(2)

iso_o country  TIME_PERIOD           gdp    gdp_pct  population  \
0   ABW   Aruba         1999  1.722905e+09  19216.199       89659   
1   ABW   Aruba         2000  1.873453e+09  20681.020       90588   

   trade_gdp_ratio    log_gdp  log_trade_gdp_ratio  log_gdp_pct  \
0        164.56000  21.267277             5.103275     9.863509   
1        145.07001  21.351049             4.977216     9.936972   

   log_population country_id  D_log_gdp_pct  D_log_gdp  
0       11.403769        ABW            NaN        NaN  
1       11.414077        ABW       0.073462   0.083773

In [549]:
attribute2000_2023= attribute2000_2023.rename(columns = {'Time': 'TIME_PERIOD'})

In [550]:
attribute2000_2023= attribute2000_2023[attribute2000_2023['TIME_PERIOD'] != 1999]

In [551]:
assert attribute2000_2023['TIME_PERIOD'].nunique() == total_long['TIME_PERIOD'].nunique() == unique_long['TIME_PERIOD'].nunique() == repro_long['TIME_PERIOD'].nunique()

* The number of countries in the long format trade datasets are 201 and the number of countries in the attribute dataset is 190. All the long format trade datasets have the same 11 more countries than the attribute dataset.
* The dyad level time invariant 'fixed' dataset has 34 countries those are not in the attribute dataset. 

* Total trade

In [552]:
remove_from_trade = list(set(total_long['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique()))

In [553]:
remove_from_trade

['PSE', 'MSR', 'WXD', 'SSD', 'TLS', 'ROU', 'TWN', 'SCG_F', 'SRB', 'MNE', 'COD']

In [554]:
total_long = total_long[~total_long['iso_o'].isin(remove_from_trade)]
total_long = total_long[~total_long['iso_d'].isin(remove_from_trade)]

In [555]:
total_long = total_long.rename(columns = {'OBS_VALUE': 'total_tr'})

In [556]:
total_long.head(3)

iso_o iso_d  TIME_PERIOD  total_tr  hysteresis_total
0   ABW   AFG         2017  0.065851          0.000000
1   ABW   AFG         2018  0.398446          0.065851
2   ABW   AFG         2019  0.254087          0.418202

* Unique cultural goods trade

In [557]:
unique_long = unique_long[~unique_long['iso_o'].isin(remove_from_trade)]
unique_long = unique_long[~unique_long['iso_d'].isin(remove_from_trade)]

In [558]:
print(unique_long['iso_o'].nunique())
print(unique_long['iso_d'].nunique())

190
190


In [559]:
unique_long = unique_long.rename(columns = {'OBS_VALUE': 'unique_cultural_tr'})

In [560]:
unique_long.head(3)

iso_o iso_d PRODUCT_HS  TIME_PERIOD  unique_cultural_tr ADJUSTMENT  \
0   ABW   AGO    HS17_97         2014            0.000028   B_ADJ_RX   
1   ABW   ARE    HS17_97         2005            0.000641   B_ADJ_RX   
2   ABW   ARE    HS17_97         2014            0.000053   B_ADJ_RX   

   hysteresis_unique  
0           0.000000  
1           0.000000  
2           0.000641

* Reproducible cultural goods trade

In [561]:
repro_long = repro_long[~repro_long['iso_o'].isin(remove_from_trade)]
repro_long = repro_long[~repro_long['iso_d'].isin(remove_from_trade)]

In [562]:
print(repro_long['iso_o'].nunique())
print(repro_long['iso_d'].nunique())

190
190


In [563]:
repro_long = repro_long.rename(columns = {'OBS_VALUE': 'reproducible_cultural_tr'})

In [564]:
repro_long.head(3)

iso_o iso_d  TIME_PERIOD  reproducible_cultural_tr              PRODUCT_HS  \
0   ABW   AFG         2019                  0.012061  ('HS17_49', 'HS17_85')   
1   ABW   AGO         2008                  0.000165            ('HS17_85',)   
2   ABW   AGO         2011                  0.000076            ('HS17_85',)   

   product_count  hysteresis_repro  
0              2          0.000000  
1              1          0.000000  
2              1          0.000165

* Fixed

In [565]:
remove_from_fixed = list(set(fixed_long['iso_o'].unique()) - set(attribute2000_2023['iso_o'].unique()))

In [566]:
fixed_long = fixed_long[~fixed_long['iso_o'].isin(remove_from_fixed)]
fixed_long = fixed_long[~fixed_long['iso_d'].isin(remove_from_fixed)]

In [567]:
print(fixed_long['iso_o'].nunique())
print(fixed_long['iso_d'].nunique())

190
190


In [568]:
#fixed_long[fixed_long['development'].isna()]

In [569]:
fixed_long.loc[fixed_long['country'] == 'Bermuda', 'development'] = 1
fixed_long.loc[fixed_long['country'] == 'Cuba', 'development'] = 0
fixed_long.loc[fixed_long['country'] == 'French Polynesia', 'development'] = 1
fixed_long.loc[fixed_long['country'] == 'Greenland', 'development'] = 1
fixed_long.loc[fixed_long['country'] == 'New Caledonia', 'development'] = 1

### panel data

In [570]:
panel_total = attribute2000_2023.merge(total_long, on = ['iso_o', 'TIME_PERIOD'])

In [571]:
print(total_long.groupby(['iso_o', 'iso_d']).ngroups)
print(panel_total.groupby(['iso_o', 'iso_d']).ngroups)

33933
33933


In [572]:
print(total_long['iso_o'].nunique(), total_long['iso_d'].nunique(), total_long['TIME_PERIOD'].nunique())
print(panel_total['iso_o'].nunique(), panel_total['iso_d'].nunique(), panel_total['TIME_PERIOD'].nunique())
print(attribute2000_2023['iso_o'].nunique(), attribute2000_2023['TIME_PERIOD'].nunique())

190 190 24
190 190 24
190 24


In [573]:
panel_total = panel_total[['country', 'iso_o', 'iso_d', 'TIME_PERIOD', 'total_tr', 'hysteresis_total', 'gdp', 'log_gdp', 'D_log_gdp', 'gdp_pct','log_gdp_pct', 'D_log_gdp_pct', 'population', 'log_population','trade_gdp_ratio', 'log_trade_gdp_ratio']]

In [574]:
panel_total.head(3)

country iso_o iso_d  TIME_PERIOD  total_tr  hysteresis_total           gdp  \
0   Aruba   ABW   ARE         2000  0.000101               0.0  1.873453e+09   
1   Aruba   ABW   ARG         2000  0.034091               0.0  1.873453e+09   
2   Aruba   ABW   ATG         2000  0.039110               0.0  1.873453e+09   

     log_gdp  D_log_gdp   gdp_pct  log_gdp_pct  D_log_gdp_pct  population  \
0  21.351049   0.083773  20681.02     9.936972       0.073462       90588   
1  21.351049   0.083773  20681.02     9.936972       0.073462       90588   
2  21.351049   0.083773  20681.02     9.936972       0.073462       90588   

   log_population  trade_gdp_ratio  log_trade_gdp_ratio  
0       11.414077        145.07001             4.977216  
1       11.414077        145.07001             4.977216  
2       11.414077        145.07001             4.977216

**Merging the previous data with unique cultural goods trade data**

In [575]:
unique_clean = unique_long.drop(columns=['PRODUCT_HS', 'ADJUSTMENT']).drop_duplicates() #removing these two columns

In [576]:
unique_clean.head(2)

iso_o iso_d  TIME_PERIOD  unique_cultural_tr  hysteresis_unique
0   ABW   AGO         2014            0.000028                0.0
1   ABW   ARE         2005            0.000641                0.0

In [577]:
panel_total2 = panel_total.merge(unique_clean, on = ['iso_o', 'iso_d', 'TIME_PERIOD'], how = 'left')

In [578]:
panel_total2.head(2)

country iso_o iso_d  TIME_PERIOD  total_tr  hysteresis_total           gdp  \
0   Aruba   ABW   ARE         2000  0.000101               0.0  1.873453e+09   
1   Aruba   ABW   ARG         2000  0.034091               0.0  1.873453e+09   

     log_gdp  D_log_gdp   gdp_pct  log_gdp_pct  D_log_gdp_pct  population  \
0  21.351049   0.083773  20681.02     9.936972       0.073462       90588   
1  21.351049   0.083773  20681.02     9.936972       0.073462       90588   

   log_population  trade_gdp_ratio  log_trade_gdp_ratio  unique_cultural_tr  \
0       11.414077        145.07001             4.977216                 NaN   
1       11.414077        145.07001             4.977216                 NaN   

   hysteresis_unique  
0                NaN  
1                NaN

In [579]:
# sanity check
panel_total2[(panel_total2['iso_o'] == 'ABW') & (panel_total2['iso_d'] == 'AGO') & (panel_total2['TIME_PERIOD'] == 2014)]

country iso_o iso_d  TIME_PERIOD  total_tr  hysteresis_total  \
1139   Aruba   ABW   AGO         2014  0.037197          4.822801   

               gdp    log_gdp  D_log_gdp   gdp_pct  log_gdp_pct  \
1139  2.790850e+09  21.749613   0.022833  26129.84    10.170834   

      D_log_gdp_pct  population  log_population  trade_gdp_ratio  \
1139       0.012177      106807       11.578779           162.88   

      log_trade_gdp_ratio  unique_cultural_tr  hysteresis_unique  
1139             5.093014            0.000028                0.0

In [580]:
panel_total2 = panel_total2[['country', 'iso_o', 'iso_d', 'TIME_PERIOD', 'total_tr', 'unique_cultural_tr', 'hysteresis_total', 'hysteresis_unique','gdp', 'log_gdp', 'D_log_gdp', 'gdp_pct','log_gdp_pct', 'D_log_gdp_pct', 'population', 'log_population','trade_gdp_ratio', 'log_trade_gdp_ratio']]

In [581]:
print(unique_long.groupby(['iso_o', 'iso_d']).ngroups)
print(panel_total2.groupby(['iso_o', 'iso_d']).ngroups)

17414
33933


In [582]:
print(unique_long['iso_o'].nunique(), unique_long['iso_d'].nunique(), unique_long['TIME_PERIOD'].nunique())
print(panel_total2['iso_o'].nunique(), panel_total2['iso_d'].nunique(), panel_total2['TIME_PERIOD'].nunique())

190 190 24
190 190 24


**Merging the previous data with reproducible goods trade data**

In [583]:
repro_long.head(2)

iso_o iso_d  TIME_PERIOD  reproducible_cultural_tr              PRODUCT_HS  \
0   ABW   AFG         2019                  0.012061  ('HS17_49', 'HS17_85')   
1   ABW   AGO         2008                  0.000165            ('HS17_85',)   

   product_count  hysteresis_repro  
0              2               0.0  
1              1               0.0

In [584]:
repro_clean = repro_long.drop(columns=['PRODUCT_HS', 'product_count']).drop_duplicates() #removing these two columns

In [585]:
panel_total3 = panel_total2.merge(repro_clean, on = ['iso_o', 'iso_d', 'TIME_PERIOD'], how='left')

In [586]:
panel_total3 = panel_total3[['country', 'iso_o', 'iso_d', 'TIME_PERIOD', 'total_tr', 'unique_cultural_tr', 'reproducible_cultural_tr', 'hysteresis_total', 'hysteresis_unique', 'hysteresis_repro','gdp', 'log_gdp', 'D_log_gdp', 'gdp_pct','log_gdp_pct', 'D_log_gdp_pct', 'population', 'log_population','trade_gdp_ratio', 'log_trade_gdp_ratio']]

In [587]:
# sanity check
panel_total3[(panel_total3['iso_o'] == 'ABW') & (panel_total3['iso_d'] == 'AGO') & (panel_total3['TIME_PERIOD'] == 2014)]

country iso_o iso_d  TIME_PERIOD  total_tr  unique_cultural_tr  \
1139   Aruba   ABW   AGO         2014  0.037197            0.000028   

      reproducible_cultural_tr  hysteresis_total  hysteresis_unique  \
1139                       NaN          4.822801                0.0   

      hysteresis_repro           gdp    log_gdp  D_log_gdp   gdp_pct  \
1139               NaN  2.790850e+09  21.749613   0.022833  26129.84   

      log_gdp_pct  D_log_gdp_pct  population  log_population  trade_gdp_ratio  \
1139    10.170834       0.012177      106807       11.578779           162.88   

      log_trade_gdp_ratio  
1139             5.093014

In [588]:
print(repro_long['iso_o'].nunique(), repro_long['iso_d'].nunique(), repro_long['TIME_PERIOD'].nunique())
print(panel_total3['iso_o'].nunique(), panel_total3['iso_d'].nunique(), panel_total3['TIME_PERIOD'].nunique())

190 190 24
190 190 24


In [589]:
rta_long = pd.read_csv("../data/cleaned/rta_long.csv")

In [590]:
rta_long.head(3)

iso_o iso_d  year  rta
0   ABW   AFG  2000    0
1   ABW   AFG  2001    0
2   ABW   AFG  2002    0

In [591]:
rta_long = rta_long.rename(columns = {'year': 'TIME_PERIOD'})

In [592]:
panel_total4 = panel_total3.merge(rta_long, on = ['iso_o', 'iso_d', 'TIME_PERIOD'], how='left')

In [593]:
panel_total4.head(2)

country iso_o iso_d  TIME_PERIOD  total_tr  unique_cultural_tr  \
0   Aruba   ABW   ARE         2000  0.000101                 NaN   
1   Aruba   ABW   ARG         2000  0.034091                 NaN   

   reproducible_cultural_tr  hysteresis_total  hysteresis_unique  \
0                       NaN               0.0                NaN   
1                  0.000104               0.0                NaN   

   hysteresis_repro  ...    log_gdp  D_log_gdp   gdp_pct  log_gdp_pct  \
0               NaN  ...  21.351049   0.083773  20681.02     9.936972   
1               0.0  ...  21.351049   0.083773  20681.02     9.936972   

   D_log_gdp_pct  population  log_population  trade_gdp_ratio  \
0       0.073462       90588       11.414077        145.07001   
1       0.073462       90588       11.414077        145.07001   

   log_trade_gdp_ratio  rta  
0             4.977216    0  
1             4.977216    0  

[2 rows x 21 columns]

In [594]:
import itertools

In [595]:
panel_total4.columns

Index(['country', 'iso_o', 'iso_d', 'TIME_PERIOD', 'total_tr',
       'unique_cultural_tr', 'reproducible_cultural_tr', 'hysteresis_total',
       'hysteresis_unique', 'hysteresis_repro', 'gdp', 'log_gdp', 'D_log_gdp',
       'gdp_pct', 'log_gdp_pct', 'D_log_gdp_pct', 'population',
       'log_population', 'trade_gdp_ratio', 'log_trade_gdp_ratio', 'rta'],
      dtype='object')

In [596]:
panel_total4_tr = panel_total4[['country', 'iso_o', 'iso_d', 'TIME_PERIOD', 'total_tr',
       'unique_cultural_tr', 'reproducible_cultural_tr', 'hysteresis_total', 'hysteresis_unique', 'hysteresis_repro', 'rta']]

In [597]:
country_list = panel_total4['iso_o'].unique()
print(len(country_list))
years = list(range(2000, 2024))
panel_df = pd.MultiIndex.from_product([country_list, country_list, years], names=['iso_o', 'iso_d', 'TIME_PERIOD'])
panel_df_blank = panel_df.to_frame(index=False)

190


In [598]:
(190*190 - 190)*24

861840

In [599]:
panel_df_blank.shape

(866400, 3)

In [600]:
panel_df_blank = panel_df_blank[panel_df_blank['iso_o'] != panel_df_blank['iso_d']]
only_trade = panel_df_blank.merge(panel_total4_tr, on = ['iso_o', 'iso_d', 'TIME_PERIOD'], how = 'left')

In [601]:
assert only_trade.shape[0] == panel_df_blank.shape[0]

In [602]:
missing_value_columns = ['total_tr', 'unique_cultural_tr', 'reproducible_cultural_tr', 'hysteresis_total', 'hysteresis_unique', 'hysteresis_repro', 'rta']

In [603]:
only_trade[missing_value_columns] = only_trade[missing_value_columns].fillna(0)

In [604]:
#print(only_trade[(only_trade['iso_o'] == 'BGD') & (only_trade['iso_d'] == 'IND')])

In [605]:
trade_time_series_attribute = only_trade.merge(attribute2000_2023, on = ['iso_o', 'TIME_PERIOD'], how = 'left')

In [606]:
assert only_trade.shape[0] == panel_df_blank.shape[0] == trade_time_series_attribute.shape[0]

In [607]:
assert attribute2000_2023['iso_o'].nunique() == only_trade['iso_o'].nunique() ==  panel_df_blank['iso_o'].nunique() == trade_time_series_attribute['iso_o'].nunique()

In [608]:
trade_time_series_attribute.head(2)

iso_o iso_d  TIME_PERIOD country_x  total_tr  unique_cultural_tr  \
0   ABW   AFG         2000       NaN       0.0                 0.0   
1   ABW   AFG         2001       NaN       0.0                 0.0   

   reproducible_cultural_tr  hysteresis_total  hysteresis_unique  \
0                       0.0               0.0                0.0   
1                       0.0               0.0                0.0   

   hysteresis_repro  ...    gdp_pct population  trade_gdp_ratio    log_gdp  \
0               0.0  ...  20681.020      90588        145.07001  21.351049   
1               0.0  ...  20740.131      91439        140.39000  21.363253   

   log_trade_gdp_ratio  log_gdp_pct  log_population  country_id  \
0             4.977216     9.936972       11.414077         ABW   
1             4.944424     9.939826       11.423428         ABW   

   D_log_gdp_pct  D_log_gdp  
0       0.073462   0.083773  
1       0.002854   0.012203  

[2 rows x 23 columns]

In [609]:
trade_time_series_attribute['D_log_gdp'] = trade_time_series_attribute['D_log_gdp'].fillna(0)
trade_time_series_attribute['D_log_gdp_pct'] = trade_time_series_attribute['D_log_gdp_pct'].fillna(0)

In [610]:
#trade_time_series_attribute[['GDP_pct', 'population', 'trade_gdp_ratio', 'log_population', 'log_trade_gdp_ratio']].isnull().mean()

In [611]:
#raise SystemExit("Stop!")

In [612]:
trade_time_series_attribute =  trade_time_series_attribute.merge(fixed_long, on = ['iso_o', 'iso_d'], how = 'left')

In [613]:
trade_time_series_attribute.columns

Index(['iso_o', 'iso_d', 'TIME_PERIOD', 'country_x', 'total_tr',
       'unique_cultural_tr', 'reproducible_cultural_tr', 'hysteresis_total',
       'hysteresis_unique', 'hysteresis_repro', 'rta', 'country_y', 'gdp',
       'gdp_pct', 'population', 'trade_gdp_ratio', 'log_gdp',
       'log_trade_gdp_ratio', 'log_gdp_pct', 'log_population', 'country_id',
       'D_log_gdp_pct', 'D_log_gdp', 'contig', 'comlang_off', 'colony',
       'comcol', 'dist', 'log_dist', 'country', 'landlocked', 'continent',
       'lat', 'lon', 'langoff_1', 'colonizer1', 'category', 'development'],
      dtype='object')

In [614]:
trade_time_series_attribute = trade_time_series_attribute.drop(columns = ['country_x', 'country_y', 'country_id'])

In [615]:
trade_time_series_attribute.head(2)

iso_o iso_d  TIME_PERIOD  total_tr  unique_cultural_tr  \
0   ABW   AFG         2000       0.0                 0.0   
1   ABW   AFG         2001       0.0                 0.0   

   reproducible_cultural_tr  hysteresis_total  hysteresis_unique  \
0                       0.0               0.0                0.0   
1                       0.0               0.0                0.0   

   hysteresis_repro  rta  ...  log_dist  country  landlocked  continent  \
0               0.0  0.0  ...  9.492342    Aruba           0    America   
1               0.0  0.0  ...  9.492342    Aruba           0    America   

     lat        lon  langoff_1  colonizer1                           category  \
0  12.55 -70.099998      Dutch         NLD  Emerging and Developing Economies   
1  12.55 -70.099998      Dutch         NLD  Emerging and Developing Economies   

   development  
0          0.0  
1          0.0  

[2 rows x 35 columns]

#### Creating variables for destination countries using a copy of the original dataset

In [616]:
partner_attributes = trade_time_series_attribute.copy()

In [617]:
partner_attributes.columns

Index(['iso_o', 'iso_d', 'TIME_PERIOD', 'total_tr', 'unique_cultural_tr',
       'reproducible_cultural_tr', 'hysteresis_total', 'hysteresis_unique',
       'hysteresis_repro', 'rta', 'gdp', 'gdp_pct', 'population',
       'trade_gdp_ratio', 'log_gdp', 'log_trade_gdp_ratio', 'log_gdp_pct',
       'log_population', 'D_log_gdp_pct', 'D_log_gdp', 'contig', 'comlang_off',
       'colony', 'comcol', 'dist', 'log_dist', 'country', 'landlocked',
       'continent', 'lat', 'lon', 'langoff_1', 'colonizer1', 'category',
       'development'],
      dtype='object')

In [618]:
#removing all dyad level factors
partner_attributes = partner_attributes.drop(columns = ['iso_d', 'total_tr', 'unique_cultural_tr', 'rta', 'reproducible_cultural_tr', 'contig', 'comlang_off',
       'colony', 'comcol', 'log_dist', 'dist', 'lat', 'lon', 'hysteresis_total', 'hysteresis_unique',
       'hysteresis_repro'])

In [619]:
partner_attributes.columns

Index(['iso_o', 'TIME_PERIOD', 'gdp', 'gdp_pct', 'population',
       'trade_gdp_ratio', 'log_gdp', 'log_trade_gdp_ratio', 'log_gdp_pct',
       'log_population', 'D_log_gdp_pct', 'D_log_gdp', 'country', 'landlocked',
       'continent', 'langoff_1', 'colonizer1', 'category', 'development'],
      dtype='object')

In [620]:
destination_attributes = {'iso_o': 'iso_d', 'country': 'country_d', 'gdp': 'gdp_d', 'log_gdp': 'log_gdp_d', 'D_log_gdp': 'D_log_gdp_d','population': 'population_d',
 'log_population': 'log_population_d', 'gdp_pct': 'gdp_pct_d', 'log_gdp_pct': 'log_gdp_pct_d', 'D_log_gdp_pct': 'D_log_gdp_pct_d','trade_gdp_ratio': 'trade_gdp_ratio_d',
'log_trade_gdp_ratio': 'log_trade_gdp_ratio_d', 'landlocked': 'landlocked_d', 'continent': 'continent_d', 'category': 'category_d','development': 'development_d',
'colonizer1': 'colonizer1_d','langoff_1': 'langoff_1_d'}

In [621]:
partner_attributes = partner_attributes.rename(columns = destination_attributes)

In [622]:
partner_attributes.head(2)

iso_d  TIME_PERIOD         gdp_d  gdp_pct_d  population_d  \
0   ABW         2000  1.873453e+09  20681.020         90588   
1   ABW         2001  1.896457e+09  20740.131         91439   

   trade_gdp_ratio_d  log_gdp_d  log_trade_gdp_ratio_d  log_gdp_pct_d  \
0          145.07001  21.351049               4.977216       9.936972   
1          140.39000  21.363253               4.944424       9.939826   

   log_population_d  D_log_gdp_pct_d  D_log_gdp_d country_d  landlocked_d  \
0         11.414077         0.073462     0.083773     Aruba             0   
1         11.423428         0.002854     0.012203     Aruba             0   

  continent_d langoff_1_d colonizer1_d                         category_d  \
0     America       Dutch          NLD  Emerging and Developing Economies   
1     America       Dutch          NLD  Emerging and Developing Economies   

   development_d  
0            0.0  
1            0.0

In [623]:
partner_attributes.columns

Index(['iso_d', 'TIME_PERIOD', 'gdp_d', 'gdp_pct_d', 'population_d',
       'trade_gdp_ratio_d', 'log_gdp_d', 'log_trade_gdp_ratio_d',
       'log_gdp_pct_d', 'log_population_d', 'D_log_gdp_pct_d', 'D_log_gdp_d',
       'country_d', 'landlocked_d', 'continent_d', 'langoff_1_d',
       'colonizer1_d', 'category_d', 'development_d'],
      dtype='object')

In [624]:
search_duplicate_row = ['iso_d', 'TIME_PERIOD']

In [625]:
partner_attributes_duplicate = partner_attributes.duplicated(subset = search_duplicate_row,  keep = 'last')

In [626]:
partner_attributes_final = partner_attributes[~partner_attributes_duplicate]

In [627]:
assert partner_attributes_final.shape[0] == attribute2000_2023.shape[0]

In [628]:
complete = trade_time_series_attribute.merge(partner_attributes_final, on = ['iso_d', 'TIME_PERIOD'], how = 'left')

In [629]:
complete.shape

(861840, 52)

In [630]:
complete.columns

Index(['iso_o', 'iso_d', 'TIME_PERIOD', 'total_tr', 'unique_cultural_tr',
       'reproducible_cultural_tr', 'hysteresis_total', 'hysteresis_unique',
       'hysteresis_repro', 'rta', 'gdp', 'gdp_pct', 'population',
       'trade_gdp_ratio', 'log_gdp', 'log_trade_gdp_ratio', 'log_gdp_pct',
       'log_population', 'D_log_gdp_pct', 'D_log_gdp', 'contig', 'comlang_off',
       'colony', 'comcol', 'dist', 'log_dist', 'country', 'landlocked',
       'continent', 'lat', 'lon', 'langoff_1', 'colonizer1', 'category',
       'development', 'gdp_d', 'gdp_pct_d', 'population_d',
       'trade_gdp_ratio_d', 'log_gdp_d', 'log_trade_gdp_ratio_d',
       'log_gdp_pct_d', 'log_population_d', 'D_log_gdp_pct_d', 'D_log_gdp_d',
       'country_d', 'landlocked_d', 'continent_d', 'langoff_1_d',
       'colonizer1_d', 'category_d', 'development_d'],
      dtype='object')

In [631]:
s = complete.columns

In [632]:
len(s)

52

In [633]:
complete[['country', 'country_d']].head(3)

country    country_d
0   Aruba  Afghanistan
1   Aruba  Afghanistan
2   Aruba  Afghanistan

In [634]:
complete['rta'].dtypes

dtype('float64')

In [635]:
complete['rta'].value_counts(dropna=False)

rta
0.0    710959
1.0    150881
Name: count, dtype: int64

In [636]:
complete = complete.fillna(0)

In [637]:
assert complete.shape[0]== only_trade.shape[0] == panel_df_blank.shape[0] == trade_time_series_attribute.shape[0] 

In [638]:
print(complete.duplicated(subset=['iso_o', 'iso_d', 'TIME_PERIOD']).any())

False


In [639]:
print(complete['iso_o'].nunique())
print(complete['iso_d'].nunique())
print(complete['TIME_PERIOD'].nunique())
print(complete['log_gdp'].nunique())
print(complete['D_log_gdp_d'].nunique())
print(complete['log_population'].nunique())
print(complete['log_population_d'].nunique())
print(complete['log_dist'].nunique())
print(complete.groupby(['iso_o', 'iso_d']).ngroups) #190*190 is 36100 possible dyads while removing 190 self loops make it (36100- 190) = 35910 dyads
print(complete.shape)

190
190
24
4527
4483
4558
4558
17927
35910
(861840, 52)


In [640]:
complete.columns

Index(['iso_o', 'iso_d', 'TIME_PERIOD', 'total_tr', 'unique_cultural_tr',
       'reproducible_cultural_tr', 'hysteresis_total', 'hysteresis_unique',
       'hysteresis_repro', 'rta', 'gdp', 'gdp_pct', 'population',
       'trade_gdp_ratio', 'log_gdp', 'log_trade_gdp_ratio', 'log_gdp_pct',
       'log_population', 'D_log_gdp_pct', 'D_log_gdp', 'contig', 'comlang_off',
       'colony', 'comcol', 'dist', 'log_dist', 'country', 'landlocked',
       'continent', 'lat', 'lon', 'langoff_1', 'colonizer1', 'category',
       'development', 'gdp_d', 'gdp_pct_d', 'population_d',
       'trade_gdp_ratio_d', 'log_gdp_d', 'log_trade_gdp_ratio_d',
       'log_gdp_pct_d', 'log_population_d', 'D_log_gdp_pct_d', 'D_log_gdp_d',
       'country_d', 'landlocked_d', 'continent_d', 'langoff_1_d',
       'colonizer1_d', 'category_d', 'development_d'],
      dtype='object')

In [641]:
# 34782/2 = 17391 #So the unique value of country pair distance column is almost near half of the number of total unique dyads in the data. 

In [642]:
complete['dyad_id'] = complete['iso_o'] + '-' + complete['iso_d'] 

In [643]:
complete.to_csv("../data/cleaned/panel_complete.csv", encoding = 'utf-8', index = False)

In [644]:
#raise SystemExit("Stop!")

### Adding node level fixed data with node attribute data sets
steps:
>1. Merging the attributeYEAR data with the nodal_fixed data for every year starting from 2000 and ending at 2023
>2. Looking for missing value in the 'category' column and manually impute 1 for developed and 0 for developing to the 'development' column based on WDI income categories. If WDI classifies the countru with hign income, then I am considering it as developed and 0 otherwise.
>3. The Stata do file produces lagged values of GDP per capita and GDP variables and adds to the long format attribute dataset. To add those lag values to the year level attributeYEAR datasets, we first rename those variables to their lowercase so that they match the stata processed dataset.
>4. Then the long format dataset is separated into 24 different dataframes and merged with the corresponding attributeYEAR dataset on the 'iso_o' column.
>5. Saving with the same name as attributeYEAR.csv, so that no few data file is created and the attribute for each year has the fixed factor columns

In [645]:
attributes = {f"at{year}": pd.read_csv(f"../data/cleaned/attribute{year}.csv") for year in range(2000, 2024)}

In [646]:
node_fix = pd.read_csv("../data/cleaned/nodal_fixed.csv")
m = set(node_fix['iso_o'].unique()) - set(at2000['iso_o'].unique())
node_fix = node_fix[~node_fix['iso_o'].isin(m)]
set(node_fix['iso_o'].unique()) - set(at2000['iso_o'].unique())
node_fix = node_fix.reset_index(drop= True)

In [647]:
node_fix.columns

Index(['iso_o', 'country', 'landlocked', 'continent', 'lat', 'lon',
       'langoff_1', 'colonizer1', 'category', 'development'],
      dtype='object')

In [648]:
dev_map = {'Bermuda': 1, 'Cuba': 0, 'French Polynesia': 1, 'Greenland': 1,'New Caledonia': 1}

In [649]:
rename_variables = {'GDP': 'gdp', 'log_GDP': 'log_gdp', 'GDP_pct': 'gdp_pct'}

In [650]:
for name, dataframe in attributes.items():
    dataframe = dataframe.merge(node_fix, on = ['iso_o'], how = 'left')
    dataframe.loc[dataframe['Country'].isin(dev_map), 'development'] = dataframe['Country'].map(dev_map)
    dataframe = dataframe.rename(columns = rename_variables)
    globals()[name] = dataframe

In [651]:
at2023.columns

Index(['Time', 'Time Code', 'Country', 'iso_o', 'gdp', 'gdp_pct', 'population',
       'trade_gdp_ratio', 'log_gdp', 'log_trade_gdp_ratio', 'country',
       'landlocked', 'continent', 'lat', 'lon', 'langoff_1', 'colonizer1',
       'category', 'development'],
      dtype='object')

In [652]:
#at2012[at2012['Country'] == 'New Caledonia']

### Adding Lag variables

In [653]:
attribute2000_2023.head(2)

iso_o country  TIME_PERIOD           gdp    gdp_pct  population  \
1   ABW   Aruba         2000  1.873453e+09  20681.020       90588   
2   ABW   Aruba         2001  1.896457e+09  20740.131       91439   

   trade_gdp_ratio    log_gdp  log_trade_gdp_ratio  log_gdp_pct  \
1        145.07001  21.351049             4.977216     9.936972   
2        140.39000  21.363253             4.944424     9.939826   

   log_population country_id  D_log_gdp_pct  D_log_gdp  
1       11.414077        ABW       0.073462   0.083773  
2       11.423428        ABW       0.002854   0.012203

In [654]:
attribute2000_2023 = attribute2000_2023[['iso_o','TIME_PERIOD', 'log_gdp_pct', 'log_population', 'D_log_gdp_pct', 'D_log_gdp']]

In [655]:
attribute2000_2023.head(2)

iso_o  TIME_PERIOD  log_gdp_pct  log_population  D_log_gdp_pct  D_log_gdp
1   ABW         2000     9.936972       11.414077       0.073462   0.083773
2   ABW         2001     9.939826       11.423428       0.002854   0.012203

In [656]:
for year, cross_section in attribute2000_2023.groupby('TIME_PERIOD'):
    globals()[f"lag{year}"] = cross_section

In [657]:
lag2023.head(3)

iso_o  TIME_PERIOD  log_gdp_pct  log_population  D_log_gdp_pct  D_log_gdp
24   ABW         2023    10.433668       11.583934       0.106236   0.106693
49   AFG         2023     6.025286       17.540113       0.146822   0.168169
74   AGO         2023     7.744799       17.419645      -0.237853  -0.207045

In [658]:
at2023.head(3)

Time Time Code      Country iso_o           gdp  gdp_pct  population  \
0  2023    YR2023  Afghanistan   AFG  1.715223e+10   413.76  41454761.0   
1  2023    YR2023      Albania   ALB  2.354718e+10  8575.17   2745972.0   
2  2023    YR2023      Algeria   DZA  2.476262e+11  5364.03  46164219.0   

   trade_gdp_ratio    log_gdp  log_trade_gdp_ratio      country  landlocked  \
0            67.58  23.565394             4.213312  Afghanistan           1   
1            82.49  23.882272             4.412677      Albania           0   
2            43.68  26.235186             3.776890      Algeria           0   

  continent        lat        lon langoff_1 colonizer1  \
0      Asia  34.516666  69.199997   Persian        NaN   
1    Europe  41.333332  19.816668  Albanian        TUR   
2    Africa  36.833332   3.000000    Arabic        FRA   

                            category  development  
0  Emerging and Developing Economies          0.0  
1  Emerging and Developing Economies          0.0  
2  Emerging and Developing Economies          0.0

In [660]:
# merging each attribute data with that years first differences data
years = range(2000, 2024)

In [675]:
merged = {}
for year in years:
    merged[year] = globals()[f"at{year}"].merge(globals()[f"lag{year}"], on = "iso_o", how = "left")

In [676]:
for year, dataframe in merged.items():
    globals()[f"at{year}_total"] = dataframe

In [679]:
at2023_total.head(3)

Time Time Code      Country iso_o           gdp  gdp_pct  population  \
0  2023    YR2023  Afghanistan   AFG  1.715223e+10   413.76  41454761.0   
1  2023    YR2023      Albania   ALB  2.354718e+10  8575.17   2745972.0   
2  2023    YR2023      Algeria   DZA  2.476262e+11  5364.03  46164219.0   

   trade_gdp_ratio    log_gdp  log_trade_gdp_ratio  ...        lon  langoff_1  \
0            67.58  23.565394             4.213312  ...  69.199997    Persian   
1            82.49  23.882272             4.412677  ...  19.816668   Albanian   
2            43.68  26.235186             3.776890  ...   3.000000     Arabic   

  colonizer1                           category  development TIME_PERIOD  \
0        NaN  Emerging and Developing Economies          0.0        2023   
1        TUR  Emerging and Developing Economies          0.0        2023   
2        FRA  Emerging and Developing Economies          0.0        2023   

  log_gdp_pct log_population  D_log_gdp_pct  D_log_gdp  
0    6.025286      17.540113       0.146822   0.168169  
1    9.056626      14.825645       0.225143   0.213661  
2    8.587471      17.647717       0.077997   0.092985  

[3 rows x 24 columns]

In [680]:
for year in range(2000, 2024):
    globals()[f"at{year}_total"].to_csv(f"../data/cleaned/attribute{year}.csv", encoding = "utf-8", index = False)